### DISCUSSION TITLE
PDF publish failure after migrating to dita-ot 4.1.2

### DISCUSSION THREAD

**Discussion by changke79** [dita-ot/dita-ot#4411] [Q&A]
We are planning to upgrade DITA-OT toolkits to DITA-OT 4.1.2 from DITA-OT 2.4.6.
After we downloaded and upgraded DITA-OT to DITA-OT 4.1.2, we run pdf publish for a simple dita 1.3 map structure which has 4 topicref children

	<topicref href="A$X000002.dita" navtitle="Chapter One" type="topic"/>
	<topicref href="A$X000003.dita" navtitle="Chapter Two" type="task"/>
	<topicref href="A$X000004.dita" navtitle="Chapter Three" type="concept"/>
	<topicref href="A$X000005.dita" navtitle="Chapter Four" type="topic"/>

Following are log information while running ant command to generate pdf document, where the same ant build file from dita-ot 2.4.6 was used.

DITA-OT 4.1.2 supports dita 2.0, and it should support earlier dita versions like dita 1.3 map, right
Is there any special instruction to follow while upgrade to new dita-ot version?
Any help would be appreciated.

 00:44:51 []   - pdf:
 00:44:51 []   - 
 00:44:51 []   - dita2pdf2.init:
 00:44:51 []   - 
 00:44:51 []   - init-properties:
 00:44:51 []   - [init-project] Initializing project
 00:44:52 []   - 
 00:44:52 []   - check-arg:
 00:44:52 []   -    [delete] Deleting directory C:\Apps\tc\tc2406\DR\Stage\DC\Ua687463d465cc534d5a9x\temp
 00:44:52 []   -     [mkdir] Created dir: C:\Apps\tc\tc2406\DR\Stage\DC\Ua687463d465cc534d5a9x\temp
 00:44:52 []   - 
 00:44:52 []   - log-arg:
 00:44:52 []   -      [echo] *****************************************************************
 00:44:52 []   -      [echo] * basedir = C:\Apps\tc\tc2406\DR\Stage\DC\Ua687463d465cc534d5a9x\publish
 00:44:52 []   -      [echo] * dita.dir = C:\Apps\tc\tc2406\DR\Module\Translators\contmgmtpublish\lib\dita-ot
 00:44:52 []   -      [echo] * transtype = pdf
 00:44:52 []   -      [echo] * tempdir = C:\Apps\tc\tc2406\DR\Stage\DC\Ua687463d465cc534d5a9x\publish\..\temp
 00:44:52 []   -      [echo] * outputdir = C:\Apps\tc\tc2406\DR\Stage\DC\Ua687463d465cc534d5a9x\publish\..\result
 00:44:52 []   -      [echo] * clean.temp = no
 00:44:52 []   -      [echo] * DITA-OT version = 4.1.2
 00:44:52 []   -      [echo] * XML parser = Xerces
 00:44:52 []   -      [echo] * XSLT processor = Saxon
 00:44:52 []   -      [echo] * collator = ICU
 00:44:52 []   -      [echo] *****************************************************************
 00:44:52 []   -      [echo] #Ant properties
 00:44:52 []   -      [echo] #Wed Feb 14 00:44:52 EST 2024
 00:44:52 []   -      [echo] dita.plugin.org.dita.base.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.dita.base
 00:44:52 []   -      [echo] dita.temp.dir=C\:\\Apps\\tc\\tc2406\\DR\\Stage\\DC\\Ua687463d465cc534d5a9x\\publish\\..\\temp
 00:44:52 []   -      [echo] args.grammar.cache=yes
 00:44:52 []   -      [echo] args.output=A$X000001
 00:44:52 []   -      [echo] dita.input.dir=C\:\\Apps\\tc\\tc2406\\DR\\Stage\\DC\\Ua687463d465cc534d5a9x\\publish
 00:44:52 []   -      [echo] dita.plugin.org.oasis-open.xdita.v0_3_0.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.oasis-open.xdita.v0_3_0
 00:44:52 []   -      [echo] dita.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot
 00:44:52 []   -      [echo] dita.plugin.org.dita.pdf2.axf.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.dita.pdf2.axf
 00:44:52 []   -      [echo] dita.empty=
 00:44:52 []   -      [echo] args.filter=ditaValueFilter.ditaval
 00:44:52 []   -      [echo] dita.resource.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\resources
 00:44:52 []   -      [echo] dita.plugin.org.oasis-open.dita.v1_3.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.oasis-open.dita.v1_3
 00:44:52 []   -      [echo] dita.plugin.org.lwdita.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.lwdita
 00:44:52 []   -      [echo] dita.plugin.org.oasis-open.dita.v2_0.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.oasis-open.dita.v2_0
 00:44:52 []   -      [echo] dita.plugin.org.dita.htmlhelp.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.dita.htmlhelp
 00:44:52 []   -      [echo] dita.output.dir=C\:\\Apps\\tc\\tc2406\\DR\\Stage\\DC\\Ua687463d465cc534d5a9x\\publish\\..\\result
 00:44:52 []   -      [echo] args.rellinks=nofamily
 00:44:52 []   -      [echo] dita.stagedir=C\:\\Apps\\tc\\tc2406\\DR\\Stage\\DC\\Ua687463d465cc534d5a9x\\publish\\..
 00:44:52 []   -      [echo] args.ditabasedir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot
 00:44:52 []   -      [echo] dita.plugin.org.dita.eclipsehelp.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.dita.eclipsehelp
 00:44:52 []   -      [echo] dita.plugin.org.dita.normalize.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.dita.normalize
 00:44:52 []   -      [echo] dita.plugin.org.dita.xhtml.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.dita.xhtml
 00:44:52 []   -      [echo] dita.plugin.org.dita.pdf2.xep.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.dita.pdf2.xep
 00:44:52 []   -      [echo] dita.plugin.org.dita.specialization.dita11.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.dita.specialization.dita11
 00:44:52 []   -      [echo] dita.plugin.com.elovirta.pdf.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\com.elovirta.pdf
 00:44:52 []   -      [echo] dita.plugin.org.oasis-open.dita.v1_2.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.oasis-open.dita.v1_2
 00:44:52 []   -      [echo] dita.plugin.org.dita.pdf2.fop.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.dita.pdf2.fop
 00:44:52 []   -      [echo] args.input=C\:\\Apps\\tc\\tc2406\\DR\\Stage\\DC\\Ua687463d465cc534d5a9x\\publish\\A$X000001.ditamap
 00:44:52 []   -      [echo] args.xml.systemid.set=yes
 00:44:52 []   -      [echo] dita.plugin.org.dita.pdf2.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.dita.pdf2
 00:44:52 []   -      [echo] dita.plugin.org.dita.html5.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.dita.html5
 00:44:52 []   -      [echo] dita.plugin.org.oasis-open.dita.techcomm.v2_0.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.oasis-open.dita.techcomm.v2_0
 00:44:52 []   -      [echo] dita.plugin.org.dita.index.dir=C\:\\Apps\\tc\\tc2406\\DR\\Module\\Translators\\contmgmtpublish\\lib\\dita-ot\\plugins\\org.dita.index
 00:44:52 []   -      [echo] dita.extname=.xml
 00:44:52 []   -      [echo] *****************************************************************
 00:44:52 []   - 
 00:44:52 []   - build-init:
 00:44:52 []   - 
 00:44:52 []   - preprocess2.init:
 00:44:52 []   -      [echo] *****************************************************************
 00:44:52 []   -      [echo] * input = C:\Apps\tc\tc2406\DR\Stage\DC\Ua687463d465cc534d5a9x\publish\A$X000001.ditamap
 00:44:52 []   -      [echo] * resources = ${args.resources}
 00:44:52 []   -      [echo] *****************************************************************
 00:44:52 []   - 
 00:44:52 []   - ditaval-merge:
 00:44:52 []   - 
 00:44:52 []   - map-reader:
 00:44:52 []   - [map-reader] Using Xerces grammar pool for DTD and schema caching.
 00:44:52 []   - [map-reader] Processing file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/publish/A$X000001.ditamap to file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/5bbed561e1ff6de0961562bb6bc8180908faf7c8.ditamap
 00:44:52 []   - [map-reader] Serializing job specification
 00:44:52 []   - 
 00:44:52 []   - map-mapref:
 00:44:53 []   -    [mapref] Processing file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/5bbed561e1ff6de0961562bb6bc8180908faf7c8.ditamap
 00:44:53 []   - 
 00:44:53 []   - map-branch-filter:
 00:44:53 []   - [branch-filter] Processing file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/5bbed561e1ff6de0961562bb6bc8180908faf7c8.ditamap
 00:44:53 []   - 
 00:44:53 []   - map-keyref:
 00:44:53 []   -    [mapref] Processing file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/5bbed561e1ff6de0961562bb6bc8180908faf7c8.ditamap
 00:44:53 []   - 
 00:44:53 []   - map-conrefpush:
 00:44:53 []   - 
 00:44:53 []   - map-conref:
 00:44:53 []   - 
 00:44:53 []   - map-profile:
 00:44:53 []   -   [profile] Processing C:\Apps\tc\tc2406\DR\Stage\DC\Ua687463d465cc534d5a9x\publish\..\temp\5bbed561e1ff6de0961562bb6bc8180908faf7c8.ditamap
 00:44:53 []   - 
 00:44:53 []   - preprocess2.maps:
 00:44:53 []   - 
 00:44:53 []   - topic-reader:
 00:44:53 []   - [topic-reader] Using Xerces grammar pool for DTD and schema caching.
 00:44:53 []   - [topic-reader] Reading file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/5bbed561e1ff6de0961562bb6bc8180908faf7c8.ditamap
 00:44:53 []   - [topic-reader] Serializing job specification
 00:44:53 []   - 
 00:44:53 []   - topic-branch-filter:
 00:44:53 []   - [branch-filter] Processing file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/5bbed561e1ff6de0961562bb6bc8180908faf7c8.ditamap
 00:44:53 []   - 
 00:44:53 []   - topic-map-conref:
 00:44:53 []   - 
 00:44:53 []   - topic-keyref:
 00:44:53 []   - 
 00:44:53 []   - topic-copy-to:
 00:44:53 []   - 
 00:44:53 []   - topic-conrefpush:
 00:44:53 []   - 
 00:44:53 []   - topic-conref:
 00:44:53 []   - 
 00:44:53 []   - topic-profile:
 00:44:53 []   - 
 00:44:53 []   - preprocess2.topic-fragment:
 00:44:53 []   - [preprocess2.topic-fragment] Processing file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/e6dbfea6fe601417ef6a493e1b7e4a71cbf3ea96.dita
 00:44:53 []   - [preprocess2.topic-fragment] Processing file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/5bbed561e1ff6de0961562bb6bc8180908faf7c8.ditamap
 00:44:53 []   - [preprocess2.topic-fragment] Failed to process XML filter: Failed to transform file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/e6dbfea6fe601417ef6a493e1b7e4a71cbf3ea96.dita: C:\Apps\tc\tc2406\DR\Stage\DC\Ua687463d465cc534d5a9x\temp\e6dbfea6fe601417ef6a493e1b7e4a71cbf3ea96.dita (The system cannot find the file specified)
 00:44:53 []   - [preprocess2.topic-fragment] Processing file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/57b110bf977b691c96289b0095b0e5ad1dd48a66.dita
 00:44:53 []   - [preprocess2.topic-fragment] Processing file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/9154d0577a6108ba11deffaa0cc394249a7f652c.dita
 00:44:53 []   - [preprocess2.topic-fragment] Failed to process XML filter: Failed to transform file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/57b110bf977b691c96289b0095b0e5ad1dd48a66.dita: C:\Apps\tc\tc2406\DR\Stage\DC\Ua687463d465cc534d5a9x\temp\57b110bf977b691c96289b0095b0e5ad1dd48a66.dita (The system cannot find the file specified)
 00:44:53 []   - [preprocess2.topic-fragment] Failed to process XML filter: Failed to transform file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/9154d0577a6108ba11deffaa0cc394249a7f652c.dita: C:\Apps\tc\tc2406\DR\Stage\DC\Ua687463d465cc534d5a9x\temp\9154d0577a6108ba11deffaa0cc394249a7f652c.dita (The system cannot find the file specified)
 00:44:53 []   - [preprocess2.topic-fragment] Processing file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/439ca3496c890eaa1dcaa8c3549de991c63ccfee.dita
 00:44:53 []   - [preprocess2.topic-fragment] Failed to process XML filter: Failed to transform file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/439ca3496c890eaa1dcaa8c3549de991c63ccfee.dita: C:\Apps\tc\tc2406\DR\Stage\DC\Ua687463d465cc534d5a9x\temp\439ca3496c890eaa1dcaa8c3549de991c63ccfee.dita (The system cannot find the file specified)
 00:44:53 []   - 
 00:44:53 []   - topic-chunk:
 00:44:53 []   - 
 00:44:53 []   - topic-move-meta-entries:
 00:44:53 []   - [move-meta] Processing file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/publish/../temp/5bbed561e1ff6de0961562bb6bc8180908faf7c8.ditamap
 00:44:53 []   - [move-meta] Loading stylesheet C:\Apps\tc\tc2406\DR\Module\Translators\contmgmtpublish\lib\dita-ot\plugins\org.dita.base\xsl\preprocess\mappull.xsl
 00:44:54 []   - [move-meta] Processing file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/5bbed561e1ff6de0961562bb6bc8180908faf7c8.ditamap
 00:44:54 []   - [move-meta] file:/C:/Apps/tc/tc2406/DR/Module/Translators/contmgmtpublish/lib/dita-ot/plugins/org.dita.base/xsl/preprocess/mappullImpl.xsl:58:9: I/O error reported by XML parser processing file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/9154d0577a6108ba11deffaa0cc394249a7f652c.dita
 00:44:54 []   - 
 00:44:54 []   - BUILD FAILED
 00:44:54 []   - C:\Apps\tc\tc2406\DR\Stage\DC\Ua687463d465cc534d5a9x\publish\build_XML0.xml:40: The following error occurred while executing this line:
 00:44:54 []   - C:\Apps\tc\tc2406\DR\Module\Translators\contmgmtpublish\lib\dita-ot\plugins\org.dita.base\build_preprocess2.xml:312: Failed to run pipeline: Failed to transform document: I/O error reported by XML parser processing file:/C:/Apps/tc/tc2406/DR/Stage/DC/Ua687463d465cc534d5a9x/temp/9154d0577a6108ba11deffaa0cc394249a7f652c.dita

**Comment by raducoravu** [comment]
Can you maybe also create a small sample DITA project on which the problem can be reproduced, zip it and attach it to this discussion?

**Reply by changke79** [reply]
Radu, Thanks! I changed to use dita command in ant build file and it worked in Windows.

**Reply by changke79** [reply]
Now I moved to Linux and got a new issue.
1. my input file name is like A$X0006.ditamap, the dita command somehow could not take this file name correctly, and reported error saying input file A.ditamap could not be located or read. It seems $ is a special character for dita-ot dita command input.
How should $ be escaped from command input?

**Reply by changke79** [reply]
the input file containing $ is not an issue on Windows. But it does not work on Linux.

**Reply by changke79** [reply]
$ is special character for environment variable replacement for Linux/Unix. Dita command interprets $X0006 as replacement of environment variable X0006. I needs to escape % character from file name. I need to do that in ant build file.

**Reply by raducoravu** [reply]
How about if you escape the $ using the \ character in the command line you are sending?
Something like "dita -i A\\$X0006.ditamap"

**Reply by changke79** [reply]
The input files are generated from our CMS system, we could change our program to escape $ from input file. 
I checked ant shell command from apache release, found that its ant shell command  escapes $, back-slash, double-quote etc, but the ant shell command bundled in dita-ot is modified and it removed the special character escaping. 
Does anyone know why dita-ot does this?
